datafusion/expr/src/udf.rs

```diff
@@ -17,12 +17,67 @@
 
 //! Udf module contains foundational types that are used to represent UDFs in DataFusion.
 
-use crate::{Expr, ReturnTypeFunction, ScalarFunctionImplementation, Signature};
+use crate::{
+    ColumnarValue, Expr, FuncMonotonicity, ReturnTypeFunction,
+    ScalarFunctionImplementation, Signature, TypeSignature, Volatility,
+};
+use arrow::array::ArrayRef;
+use arrow::datatypes::DataType;
+use datafusion_common::{internal_err, DataFusionError, Result};
+use std::any::Any;
 use std::fmt;
 use std::fmt::Debug;
 use std::fmt::Formatter;
 use std::sync::Arc;
 
+// TODO(PR): add doc comments
+pub trait ScalarFunctionDef: Any + Sync + Send + std::fmt::Debug {
+    /// Return as [`Any`] so that it can be
+    /// downcast to a specific implementation.
+    fn as_any(&self) -> &dyn Any;
+
+    // May return 1 or more name as aliasing
+    fn name(&self) -> &[&str];
+
+    fn input_type(&self) -> TypeSignature;
+
+    fn return_type(&self) -> FunctionReturnType;
+
+    fn execute(&self, _args: &[ArrayRef]) -> Result<ArrayRef> {
+        internal_err!("This method should be implemented if `supports_execute_raw()` returns `false`")
+    }
+
+    fn volatility(&self) -> Volatility;
+
+    fn monotonicity(&self) -> Option<FuncMonotonicity>;
+
+    // ===============================
+    // OPTIONAL METHODS START BELOW
+    // ===============================
+
+    /// `execute()` and `execute_raw()` are two possible alternative for function definition:
+    /// If returns `false`, `execute()` will be used for execution;
+    /// If returns `true`, `execute_raw()` will be called.
+    fn use_execute_raw_instead(&self) -> bool {
```

Comment:
Rational for this:
built-in functions now have two kinds of implementation
1. `execute()` -- It's a more common and easy-to-implement interface for UDFs, and can be converted to the more general `execute_raw()` case using `make_scalar_function()`
2. `execute_raw()` -- Fewer existing functions are directly implemented using this interface

Though a single `execute_raw()` can cover all existing cases, this design can make the general case easier to implement for UDFs

Reply:
I agree we should make the initial implementation concise